### PR TITLE
Shutdown Lock Service & Background Sweeper on TM.close()

### LINKED ISSUE #2099: Possible Thread Leak in TransactionManagers.close()
```
    @Test
    public void buildAllTheTMs() {
        AtlasDbConfig realConfig = ImmutableAtlasDbConfig.builder()
                .keyValueService(new InMemoryAtlasDbConfig())
                .defaultLockTimeoutSeconds(120)
                .build();
        for (int i = 0; i < 1000000; i++) {
            TransactionManager mgr = TransactionManagers.create(realConfig, Optional::empty, ImmutableSet.of(), environment, false);
            Uninterruptibles.sleepUninterruptibly(100, TimeUnit.MILLISECONDS);
            mgr.close();
        }
    }
```

You'd expect this to be fine, but it actually results in an OOM ("unable to create native thread"). 

@chrisalice

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**: Fix #2099 

**Implementation Description (bullets)**:
- Close the lock service on shutdown, provided it is closeable
- Register a callback that invokes `BackgroundSweeperImpl.shutdown()` when a transaction manager is closed.

**Concerns (what feedback would you like?)**:
- Are the concurrency semantics sound?
- Is there an easier way to shut down the background sweeper when the TM is closed?
- Is there a good way of preventing users from registering arbitrary callbacks (would prefer not to support this)? Seems a bit weird that users get a concrete SerializableTM rather than just the TM interface. Maybe I should create a fresh `CleanupTask` interface (unrelated to the cleaner's cleanup tasks), though that seems like just an exercise in naming.

**Where should we start reviewing?**: `SnapshotTransactionManager.java`

**Priority (whenever / two weeks / yesterday)**: whenever. Most users don't typically open and close transaction managers, though internal security product does.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2102)

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java [hunk 1]
```diff
@@ -218,11 +228,47 @@ protected SnapshotTransaction createTransaction(
...
+            keyValueService.close();
+            closeLockServiceIfPossible();
+            for (Runnable callback : closingCallbacks) {
+                callback.run();
+            }
+        }
+    }
+
+    private void closeLockServiceIfPossible() {
+        if (lockService instanceof Closeable) {
```
Comment (chrisalice) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java]: AutoCloseable?

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManagerTest.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        assertThatThrownBy(() -> snapshotTransactionManager.registerClosingCallback(null))
+                .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    public void invokesCallbackOnClose() {
+        Runnable callback = mock(Runnable.class);
+
+        snapshotTransactionManager.registerClosingCallback(callback);
+        snapshotTransactionManager.close();
```
Comment (chrisalice) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManagerTest.java]: nit: `verify(callback, never()).run()` before this

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        verify(callback, times(1)).run();
```
Comment (chrisalice) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManagerTest.java]: `times(1)` is the default, you can leave it off

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    public void invokesMultipleSuccessfulCallbacksInOrder() {
+        Runnable callback1 = mock(Runnable.class);
+        Runnable callback2 = mock(Runnable.class);
+        InOrder inOrder = inOrder(callback1, callback2);
+
+        snapshotTransactionManager.registerClosingCallback(callback1);
+        snapshotTransactionManager.registerClosingCallback(callback2);
+        snapshotTransactionManager.close();
+        inOrder.verify(callback1).run();
+        inOrder.verify(callback2).run();
```
Comment (chrisalice) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManagerTest.java]: Other order?

File: lock-api/src/main/java/com/palantir/lock/client/LockRefreshingRemoteLockService.java [hunk 5]
```diff
@@ -131,4 +132,10 @@ public void dispose() {
         exec.shutdown();
         isClosed = true;
     }
+
+    @Override
+    public void close() throws IOException {
+        super.close();
+        dispose();
```
Comment (chrisalice) [inline lock-api/src/main/java/com/palantir/lock/client/LockRefreshingRemoteLockService.java]: Is it safe to call this multiple times?

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManagerTest.java (hunk 2 above)
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManagerTest.java]: yep, added.

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManagerTest.java (hunk 3 above)
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManagerTest.java]: Yep! (I left them out in 112-113, even, but forgot you could do that here)

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManagerTest.java (hunk 4 above)
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManagerTest.java]: Hmm, yep now that you mention it close()s usually work in reverse order, so going in reverse order of registration seems to be more expected here

File: lock-api/src/main/java/com/palantir/lock/client/LockRefreshingRemoteLockService.java (hunk 5 above)
Comment (jeremyk-91) [inline lock-api/src/main/java/com/palantir/lock/client/LockRefreshingRemoteLockService.java]: I checked: yep. ExecutorService shutdown() is idempotent, as is writing the false to `isClosed`.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java (hunk 1 above)
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java]: Yep, makes sense. As far as I know all actual implementations of the lock service use regular `Closeable`, but no reason we should fail to close something that's `AutoCloseable` but not `Closeable`

File: lock-api/src/main/java/com/palantir/lock/ForwardingRemoteLockService.java [hunk 6]
```diff
@@ -15,11 +15,13 @@
...
 package com.palantir.lock;
 
+import java.io.Closeable;
+import java.io.IOException;
 import java.util.Set;
 
 import com.google.common.collect.ForwardingObject;
 
-public abstract class ForwardingRemoteLockService extends ForwardingObject implements RemoteLockService {
+public abstract class ForwardingRemoteLockService extends ForwardingObject implements CloseableRemoteLockService {
```
Comment (hsaraogi) [inline lock-api/src/main/java/com/palantir/lock/ForwardingRemoteLockService.java]: I guess `ForwardingLockService` should also implement the closeable.

File: lock-api/src/main/java/com/palantir/lock/client/LockRefreshingRemoteLockService.java [hunk 7]
```diff
@@ -131,4 +132,10 @@ public void dispose() {
         exec.shutdown();
         isClosed = true;
     }
+
+    @Override
+    public void close() throws IOException {
```
Comment (hsaraogi) [inline lock-api/src/main/java/com/palantir/lock/client/LockRefreshingRemoteLockService.java]: Add to `LockRefreshingLockService` as well.

File: lock-api/src/main/java/com/palantir/lock/ForwardingRemoteLockService.java (hunk 6 above)
Comment (jeremyk-91) [inline lock-api/src/main/java/com/palantir/lock/ForwardingRemoteLockService.java]: Yep, wired that up to close the delegate.

File: lock-api/src/main/java/com/palantir/lock/client/LockRefreshingRemoteLockService.java (hunk 7 above)
Comment (jeremyk-91) [inline lock-api/src/main/java/com/palantir/lock/client/LockRefreshingRemoteLockService.java]: Alright, added.

**Pull request merged**